heron/stmgr/src/cpp/manager/stateful-restorer.cpp

```diff
@@ -0,0 +1,260 @@
+  for (auto task_id : get_ckpt_pending_) {
+    ckptmgr_->GetInstanceState(*(server_->GetInstanceInfo(task_id)), checkpoint_id_);
+    multi_count_metric_->scope(METRIC_CKPT_REQUESTS)->incr();
+  }
+}
+
+void StatefulRestorer::HandleCheckpointState(proto::system::StatusCode _status, sp_int32 _task_id,
+                                       sp_string _checkpoint_id,
+                                       const proto::ckptmgr::InstanceStateCheckpoint& _state) {
+  LOG(INFO) << "Got InstanceState from checkpoint mgr for task " << _task_id
+            << " and checkpoint " << _state.checkpoint_id();
+  multi_count_metric_->scope(METRIC_CKPT_RESPONSES)->incr();
+  if (!in_progress_) {
+    LOG(INFO) << "Ignoring it because we are not in restore";
+    multi_count_metric_->scope(METRIC_CKPT_RESPONSES_IGNORED)->incr();
+    return;
+  }
+  if (_checkpoint_id != checkpoint_id_) {
+    LOG(INFO) << "InstanceState from checkpont mgr for a checkpoint_id "
+              << _checkpoint_id << " that is different from ours "
+              << checkpoint_id_;
+    multi_count_metric_->scope(METRIC_CKPT_RESPONSES_IGNORED)->incr();
+    return;
+  }
+  if (_status == proto::system::OK) {
+    if (_state.checkpoint_id() != checkpoint_id_) {
+      LOG(WARNING) << "Discarding state retrieved from checkpoint mgr because the checkpoint"
+                   << " id in the response does not match ours " << checkpoint_id_;
+      multi_count_metric_->scope(METRIC_CKPT_RESPONSES_IGNORED)->incr();
+      return;
+    }
+    if (server_->SendRestoreInstanceStateRequest(_task_id, _state)) {
+      multi_count_metric_->scope(METRIC_INSTANCE_RESTORE_REQUESTS)->incr();
+      get_ckpt_pending_.erase(_task_id);
+    }
+    // Note that if we are unable to send the restore instance state request(i.e. if the
+    // last call returns false, its because of connection issues and we will get back to
+    // restore anyways.
+  } else {
+    LOG(INFO) << "InstanceState from checkpont mgr contained non ok status " << _status;
+    in_progress_ = false;
+    time_spent_metric_->Stop();
+    multi_count_metric_->scope(METRIC_CKPT_RESPONSES_ERROR)->incr();
+    multi_count_metric_->scope(METRIC_START_RESTORE_FAILED)->incr();
+    restore_done_watcher_(_status, checkpoint_id_, restore_txid_);
+  }
+}
+
+void StatefulRestorer::HandleInstanceRestoredState(sp_int32 _task_id,
+                                                   const proto::system::StatusCode _status,
+                                                   const std::string& _checkpoint_id) {
+  LOG(INFO) << "Instance " << _task_id << " restored its state for " << _checkpoint_id
+            << " with status " << _status;
+
+  // We don't handle well if instances are not able to restore
+  // TODO(skukarni) Fix this
+  CHECK_EQ(_status, proto::system::OK);
+
+  multi_count_metric_->scope(METRIC_INSTANCE_RESTORE_RESPONSES)->incr();
+  if (!in_progress_) {
+    LOG(INFO) << "Ignoring the message because we are not in Restore";
```

Comment:
could you include the _checkpoint_id in this log message. also is this an unexpected condition (i.e., should it be a warning)?